sklearn/gaussian_process/kernels.py

```diff
@@ -0,0 +1,940 @@
+""" Kernels for Gaussian process regression and classification.
+
+The kernels in this module allow kernel-engineering, i.e., they can be
+combined via the "+" and "*" operators. These expressions can also contain
+scalar values, which are automatically converted to a constant kernel.
+
+All kernels allow (analytic) gradient-based hyperparameter optimization.
+The space of hyperparameters can be specified by giving lower und upper
+boundaries for the value of each hyperparameter (the search space is thus
+rectangular). This can be achieved by using a pair or triple instead of a
+single float wherever a parameter value is specified. In case of a pair,
+the first value specifies the lower boundary and the second value the upper
+boundary. In case of a triple, the middle value specified the initial value
+of the parameter during hyperparameter-optimization.
+"""
+
+# Author: Jan Hendrik Metzen <jhm@informatik.uni-bremen.de>
+# Licence: BSD 3 clause
+
+# Note: this module is strongly inspired by the kernel module of the george
+#       package.
+
+from abc import ABCMeta, abstractmethod
+from functools import partial
+import inspect
+
+import numpy as np
+from scipy.spatial.distance import pdist, cdist, squareform
+
+from ..metrics.pairwise import pairwise_kernels
+from ..externals import six
+from ..base import clone
+
+
+class Kernel(six.with_metaclass(ABCMeta)):
+    """ Base class for all kernels."""
+
+    def __init__(self, theta=1.0, thetaL=1e-5, thetaU=np.inf):
+        if not np.iterable(theta):
+            theta = np.array([theta])
+        self.theta = np.asarray(theta, dtype=np.float)
+        self.bounds = (np.asarray(thetaL, dtype=np.float),
+                       np.asarray(thetaU, dtype=np.float))
+
+    def get_params(self, deep=True):
+        """Get parameters of this kernel.
+
+        Parameters
+        ----------
+        deep: boolean, optional
+            If True, will return the parameters for this estimator and
+            contained subobjects that are estimators.
+
+        Returns
+        -------
+        params : mapping of string to any
+            Parameter names mapped to their values.
+        """
+        params = dict(theta=self.theta, thetaL=self.bounds[:, 0],
```

Comment:
> We could define the default parameters to be optimized at class level but allow fixing some of the hyperparameters via an optional parameter of __init__

It seems sensible. On the other hand, it could also very legitimately be the optimizer who is given charge of this. I don't know.

> `__init__(sigma_1, sigma_2, sigma_1_lower=None, sigma_1_upper=None, sigma_2_lower=None, sigma_2_upper=None)`

I would go for the second of your options here, always keeping in mind that the bounds will most probably be set in a non-positional way. I added the default values `None`. I was thinking that by not specifying bounds one could implicitly state that one does not want to optimize. Then again, that is very implicit and the fact that one doesn't set bounds may not necessarily imply  that one does not want to optimize.

My complexity alarm bells are starting to ring here, but I have no good suggestion right now as to how to tone it down.